In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [2]:
churnData = pd.read_csv('Customer-Churn.csv')

In [3]:
churnData

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

     OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0                No          Yes               No          No          No   
1               Yes           No              Yes          No          No   
2               Yes          Yes               No          No          No   
3               Yes           No              Yes         Yes          No   
4                No           No               No          No          No   
...             ...          ...              ...         ...         ...   
7038            Yes           No              Yes         Yes         Yes   
7039             No          Yes              Yes          No         Yes   
7040            Yes           No               No          No          No   
7041             No           No               No          No          No   
7042            Yes           No              Yes         Yes         Yes   

     StreamingMovies        Contract  MonthlyCharges TotalCharges Churn  
0                 No  Month-to-month           29.85        29.85    No  
1                 No        One year           56.95       1889.5    No  
2                 No  Month-to-month           53.85       108.15   Yes  
3                 No        One year           42.30      1840.75    No  
4                 No  Month-to-month           70.70       151.65   Yes  
...              ...             ...             ...          ...   ...  
7038             Yes        One year           84.80       1990.5    No  
7039             Yes        One year          103.20       7362.9    No  
7040              No  Month-to-month           29.60       346.45    No  
7041              No  Month-to-month           74.40        306.6   Yes  
7042             Yes        Two year          105.65       6844.5    No  

[7043 rows x 16 columns]

In [4]:
churnData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7043 non-null   object 
 15  Churn             7043 non-null   object 
dtypes: float64(1), int64(2), object(13)
memory

In [5]:
churnData['TotalCharges']=pd.to_numeric(churnData['TotalCharges'],errors='coerce')

In [6]:
churnData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7032 non-null   float64
 15  Churn             7043 non-null   object 
dtypes: float64(2), int64(2), object(12)
memory

In [7]:
churnData['TotalCharges'] = churnData['TotalCharges'].fillna(np.mean(churnData['TotalCharges']))

In [8]:
churnData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7043 non-null   float64
 15  Churn             7043 non-null   object 
dtypes: float64(2), int64(2), object(12)
memory

In [9]:
numerical=churnData.select_dtypes(exclude=object)

In [10]:
numerical

SeniorCitizen  tenure  MonthlyCharges  TotalCharges
0                 0       1           29.85         29.85
1                 0      34           56.95       1889.50
2                 0       2           53.85        108.15
3                 0      45           42.30       1840.75
4                 0       2           70.70        151.65
...             ...     ...             ...           ...
7038              0      24           84.80       1990.50
7039              0      72          103.20       7362.90
7040              0      11           29.60        346.45
7041              1       4           74.40        306.60
7042              0      66          105.65       6844.50

[7043 rows x 4 columns]

In [11]:
churnData1 = pd.concat((numerical,churnData["Churn"]),axis=1)

In [12]:
churnData1

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
0                 0       1           29.85         29.85    No
1                 0      34           56.95       1889.50    No
2                 0       2           53.85        108.15   Yes
3                 0      45           42.30       1840.75    No
4                 0       2           70.70        151.65   Yes
...             ...     ...             ...           ...   ...
7038              0      24           84.80       1990.50    No
7039              0      72          103.20       7362.90    No
7040              0      11           29.60        346.45    No
7041              1       4           74.40        306.60   Yes
7042              0      66          105.65       6844.50    No

[7043 rows x 5 columns]

In [13]:
churnData1.dtypes

SeniorCitizen       int64
tenure              int64
MonthlyCharges    float64
TotalCharges      float64
Churn              object
dtype: object

In [14]:
X =churnData1 .drop(['Churn'], axis=1)
y =churnData1['Churn']

In [15]:
churnData1['Churn'].value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

In [16]:
from sklearn.utils import resample
category_0 = churnData1[churnData1['Churn'] == 'No']
category_1 = churnData1[churnData1['Churn'] == 'Yes']

In [17]:
category_0

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
0                 0       1           29.85         29.85    No
1                 0      34           56.95       1889.50    No
3                 0      45           42.30       1840.75    No
6                 0      22           89.10       1949.40    No
7                 0      10           29.75        301.90    No
...             ...     ...             ...           ...   ...
7037              0      72           21.15       1419.40    No
7038              0      24           84.80       1990.50    No
7039              0      72          103.20       7362.90    No
7040              0      11           29.60        346.45    No
7042              0      66          105.65       6844.50    No

[5174 rows x 5 columns]

In [18]:
category_0_undersampled = resample(category_0, 
                                   replace=False, 
                                   n_samples = len(category_1))

In [19]:
category_0_undersampled

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
1902              0      72           90.60       6441.85    No
6621              0      46           19.40        958.15    No
2256              0      15           65.60       1010.00    No
6510              0      46           99.65       4630.20    No
1501              0      13           19.95        243.65    No
...             ...     ...             ...           ...   ...
411               0      46           24.45       1066.15    No
1678              1      67          105.40       7035.60    No
6531              0      25           65.80       1679.65    No
3212              0      40           80.00       3168.75    No
813               0      34           30.40        938.65    No

[1869 rows x 5 columns]

In [20]:
data_downsampled = pd.concat([category_0_undersampled, category_1], axis=0)

In [21]:
data_downsampled 

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
1902              0      72           90.60       6441.85    No
6621              0      46           19.40        958.15    No
2256              0      15           65.60       1010.00    No
6510              0      46           99.65       4630.20    No
1501              0      13           19.95        243.65    No
...             ...     ...             ...           ...   ...
7021              0      12           59.80        727.80   Yes
7026              0       9           44.20        403.35   Yes
7032              1       1           75.75         75.75   Yes
7034              0      67          102.95       6886.25   Yes
7041              1       4           74.40        306.60   Yes

[3738 rows x 5 columns]

In [22]:
data_downsampled['Churn'].value_counts()

Yes    1869
No     1869
Name: Churn, dtype: int64

In [23]:
category_1_oversampled = resample(category_1, 
                                  replace=True, 
                                  n_samples = len(category_0))

In [24]:
category_1_oversampled

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
4682              1      27           88.30       2467.75   Yes
6368              1       2           94.00        181.70   Yes
358               1      27           30.75        805.10   Yes
2543              0      13           98.15       1230.25   Yes
3600              1      51           84.20       4146.05   Yes
...             ...     ...             ...           ...   ...
5482              0      33           73.90       2405.05   Yes
6030              0      43           49.05       2076.20   Yes
2184              1       2           88.55        179.25   Yes
353               0       4           84.80        371.90   Yes
3498              0      43           75.20       3198.60   Yes

[5174 rows x 5 columns]

In [25]:
data_upsampled = pd.concat([category_0, category_1_oversampled], axis=0)

In [26]:
data_upsampled

SeniorCitizen  tenure  MonthlyCharges  TotalCharges Churn
0                 0       1           29.85         29.85    No
1                 0      34           56.95       1889.50    No
3                 0      45           42.30       1840.75    No
6                 0      22           89.10       1949.40    No
7                 0      10           29.75        301.90    No
...             ...     ...             ...           ...   ...
5482              0      33           73.90       2405.05   Yes
6030              0      43           49.05       2076.20   Yes
2184              1       2           88.55        179.25   Yes
353               0       4           84.80        371.90   Yes
3498              0      43           75.20       3198.60   Yes

[10348 rows x 5 columns]

In [27]:
data_upsampled['Churn'].value_counts()

No     5174
Yes    5174
Name: Churn, dtype: int64

In [28]:
X1=data_upsampled.drop(['Churn'],axis=1)
y1=data_upsampled['Churn']

In [29]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X1,y1, random_state = 0)

In [30]:
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(max_depth=3)
model.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=3)

In [31]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(max_depth=5,
                             min_samples_split=20,
                             min_samples_leaf =20)
clf.fit(X_train, y_train)
print("Train score: ", clf.score(X_train, y_train))
print("Test Score: ",clf.score(X_test, y_test))

Train score:  0.7593093673495683
Test Score:  0.7460378817162737


In [33]:
pred = clf.predict(X_test)

In [34]:
pred

array(['No', 'Yes', 'No', ..., 'Yes', 'No', 'No'], dtype=object)

In [37]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,pred)

array([[942, 323],
       [334, 988]])

In [38]:

from sklearn.model_selection import cross_val_score
clf = RandomForestClassifier(max_depth=5,
                             min_samples_split=20,
                             min_samples_leaf =20)
cross_val_scores = cross_val_score(clf, X_train, y_train, cv=10)
print("Mean cross validation score: ",np.mean(cross_val_scores))

Mean cross validation score:  0.7535092345659357


In [42]:
from sklearn.model_selection import RandomizedSearchCV
max_depth_choices= [3,4,5,6,7,8,9,10,None]
criterion_choices = ['gini','entropy']
min_samples_split_choices = [2,3,4,5,6,7,8,9,10]
min_samples_leaf_choices = [2,3,4,5,6,7,8,9,10]
max_features_choices = [2,3,4,5,6]
random_grid = {'max_depth': max_depth_choices,
               'criterion': criterion_choices,
               'min_samples_split': min_samples_split_choices,
               'min_samples_leaf': min_samples_leaf_choices,
               'max_features': max_features_choices}

In [43]:
model = DecisionTreeClassifier()
random_search = RandomizedSearchCV(estimator = model, param_distributions = random_grid, n_iter=100, cv = 5, n_jobs = 10)


In [44]:
random_search

RandomizedSearchCV(cv=5, estimator=DecisionTreeClassifier(), n_iter=100,
                   n_jobs=10,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [3, 4, 5, 6, 7, 8, 9, 10,
                                                      None],
                                        'max_features': [2, 3, 4, 5, 6],
                                        'min_samples_leaf': [2, 3, 4, 5, 6, 7,
                                                             8, 9, 10],
                                        'min_samples_split': [2, 3, 4, 5, 6, 7,
                                                              8, 9, 10]})

In [47]:
random_search.fit(X_train,y_train)

/Users/neemahelan/opt/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_search.py:918: UserWarning: One or more of the test scores are non-finite: [0.75016156 0.76420671 0.76691007        nan 0.74243043 0.75905182
        nan 0.72271692        nan 0.74358906        nan        nan
 0.76214502        nan        nan 0.75389726 0.72181428        nan
        nan        nan        nan        nan 0.72426149        nan
 0.74101183 0.75879359 0.72271692        nan 0.76253037 0.75016065
 0.74990433 0.78443443 0.72374611 0.7430741  0.72374611 0.75634646
        nan        nan        nan 0.76639635        nan        nan
 0.75286716 0.76922875        nan        nan        nan        nan
 0.74874478 0.76046876 0.75364102 0.74449046        nan 0.75943817
 0.75583175        nan 0.7455218         nan        nan        nan
 0.74848556 0.71382559        nan 0.74835686        nan        nan
 0.74951557 0.75415574        nan        nan        nan        nan
        nan 0.71936666 0.75338205 0.

RandomizedSearchCV(cv=5, estimator=DecisionTreeClassifier(), n_iter=100,
                   n_jobs=10,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [3, 4, 5, 6, 7, 8, 9, 10,
                                                      None],
                                        'max_features': [2, 3, 4, 5, 6],
                                        'min_samples_leaf': [2, 3, 4, 5, 6, 7,
                                                             8, 9, 10],
                                        'min_samples_split': [2, 3, 4, 5, 6, 7,
                                                              8, 9, 10]})

In [48]:
random_search.best_params_

{'min_samples_split': 9,
 'min_samples_leaf': 2,
 'max_features': 4,
 'max_depth': None,
 'criterion': 'gini'}

In [49]:
random_search.best_score_

0.7867547679582584

In [51]:
pred = random_search.predict(X_test)

In [52]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,pred)

array([[ 969,  296],
       [ 225, 1097]])

False positive =335
False negative =303 
Precision=0.78
Recall=0.82

We can see that the model correctly predicted “No” 969 times, and incorrectly predicted “No” when the actual churn was “Yes” 296 times.
Likewise, the model correctly predicted “Yes” when the actual answer was “Yes” 1097 times. At the same time, it incorrectly predicted “Yes” when the actual churn was “No” 225 times.
Our overall accuracy is 79%. A simple baseline model is to predict the majority class, which in our case is “Yes”.

If we change the max_depth,min_samples_split and min_samples_leaf under the cross validation we may get good accuracy